docs/actions-scripts.md

```diff
@@ -1,23 +1,103 @@
+
 * **Avoid code that consumes a lot of time or memory.** The server is not an infinite resource. We can advise on code optimisation if run-times become problematic.  A good strategy is to split your processing into separate project pipeline actions; the job runner can then choose to run them in parallel if sufficient resources are available.
-* **Write code that runs across different platforms.**
-Since code will be run both locally and within a Linux-based Docker environment. In practice, this just means ensuring you use forward-slashes `/` for directories.
+
+* **Write code that runs in the OpenSAFELY platform.**
+Code will be run within a Linux-based Docker environment. In practice, this just means ensuring you use forward-slashes `/` for directories.
+
 * **Structure your code into discrete chunks, both within scripts, and by splitting into different pipeline actions.**
 This helps with:
 	* readability
 	* bug-finding
 	* parallelisation via the project pipeline
 
 
+## Reading and Writing Outputs
+
+Scripted actions can read and write output files that are saved in the workspace. These generally fall into two categories: large files of `highly_sensitive` data to feed into other actions, or smaller `moderately_sensitive` outputs intended for review and release.
+
+
+### Large `highly_sensitive` Output Files
+
+It is important that the right files formats are used for large data files. The wrong formats can waste disk space, execution time, and server memory. The specific formats used vary with language ecosystem, but they should always be compressed.
+
+Note: the `cohortextractor` command produces `csv.gz` outputs, which is the current recommended output to be consumed by all supported languages.
+
+
+=== "Python"
+
+    ```python
+    # read compressed csv output from cohortextractor
+    pd.read_csv("output/input.csv.gz)
+
+    # write compressed feather file
+    df.to_feather("output/model.feather", compression="zstd")
+
+    # read feather file, decompressed automatically
+    pd.read_feather("output/input.feather")
+    ```
+
+=== "R"
+
+    ```r
+    # read compressed csv output from cohortextractor
+    df <- readr::read_csv("output/input.csv.gz")
+
+    # write a compressed feather file
+    arrow::write_feather(df, "output/model.feather", compression = "zstd")
+
+    # read a feather file, decompressed automatically
+    df <- arrow::read_feather("output/input.feather")
+    ```
+
+=== "Stata"
+
+    ```stata
+    // stata cannot handle compressed csv files directly, so unzip first to a plain csv file
+    // the unzipped file will be discarded when the action finishes.
+    !gunzip output/input.csv.gz
+    // now import the uncompressed csv using delimited
+    import delimited using output/input.csv
+
+    // save in compressed dta.gz format
+    gzsave output/model.dta.gz
+
+    // load a compressed .dta.gz file
+    gzload output/input.dta.gz
+
+    ```
```

Comment:
thought: This is all relatively simple and looks plausible. I don't know R or Stata really, so there may be something I've missed.